### PR TITLE
chore: apisix http client

### PR DESCRIPTION
**Pull request by tokers**
Add package `pkg/apisix`.

Add Interface `apisix.APISIX` as the unique APISIX client interface.

The simple use example:

```go
cli, err := apisix.NewForOptions(&apisix.ClusterOptions{
    Name: "default",
    BaseURL: "http://127.0.0.1:9080/apisix/admin",
    AdminKey: "0x123fff"
})
if err != nil {
    log.Errorf("failed to create apisix client: %s", err)
    return
}

// Resource operations.
cli.Cluster("default").Route().List()
cli.Cluster("default").Route().Create(ctx, route)
cli.Cluster("default").Route().Update(ctx, route)
cli.Cluster("default").Route().Delete(ctx, route)

// Cluster oerations
cli.AddCluster(clusterOptions)
cli.ListClusters()
```

### CODE REVIEW

File: pkg/apisix/resource_test.go [hunk 1]
```diff
@@ -21,15 +21,15 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func TestServiceUnmarshalJSON(t *testing.T) {
-	var svc Services
+func TestItemUnmarshalJSON(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (tokers) [reply pkg/apisix/resource_test.go:24]: All resources will use this function, so just test itself.

File: pkg/apisix/stub.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+	var cr createResponse
+	dec := json.NewDecoder(resp.Body)
+	if err := dec.Decode(&cr); err != nil {
+		return nil, err
+	}
+	return &cr, nil
+}
+
+func (s *stub) updateResource(ctx context.Context, url string, body io.Reader) error {
```
[the remark this answers is not in the document]
Comment (tokers) [reply pkg/apisix/stub.go:94]: OK, will add it.

File: pkg/apisix/resource.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+	ServiceId  *string                `json:"service_id"`
+	Host       *string                `json:"host"`
+	URI        *string                `json:"uri"`
+	Desc       *string                `json:"desc"`
+	Methods    []*string              `json:"methods"`
+	Plugins    map[string]interface{} `json:"plugins"`
+}
+
+// route decodes item.Value and converts it to v1.Route.
+func (i *item) route(group string) (*v1.Route, error) {
```
[the remark this answers is not in the document]
Comment (tokers) [reply pkg/apisix/resource.go:81]: We only need one base url to specify the manager api, it's manager api's responsibility to organize the APISIX cluster.

[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/apisix/resource.go:83]: Can this PR be merge? If use lua admin api , the logic is different.

[the remark this answers is not in the document]
Comment (tokers) [reply pkg/apisix/resource.go:83]: Even if the Lua admin api is in use, after the ingress controller is capable to filter namespaces, the multi cluster can be supported by deploying multiple ingress controller instances in different namespaces (with their corresponding APISIX instances) and watches their own namespace. That's more concise, from design's point of view.

File: pkg/apisix/stub.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tokers) [reply pkg/apisix/stub.go]: Solved.

File: pkg/apisix/resource.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tokers) [reply pkg/apisix/resource.go]: Multiple clusters supported.

File: pkg/ingress/controller/controller.go [hunk 4]
```diff
@@ -58,7 +60,16 @@ func NewController(cfg *config.Config) (*Controller, error) {
 		podNamespace = "default"
 	}
 
-	conf.SetBaseUrl(cfg.APISIX.BaseURL)
+	client, err := apisix.NewForOptions(&apisix.ClusterOptions{
+		Name:     "",
```
Comment (gxthrj) [inline pkg/ingress/controller/controller.go:64]: DefaultClusterName is “” ？

**Review by gxthrj — APPROVED** [review]

File: pkg/ingress/controller/controller.go (hunk 4 above)
Comment (tokers) [reply pkg/ingress/controller/controller.go:64]: It's consistent with the current design, i.e. group is "".

**Pull request merged**